### ISSUE TITLE
Springboot 1.5.1 Upgrade - MVC Exception Handling Issues

### ISSUE THREAD

**Issue by ashwgupt** [spring-projects/spring-boot#8416]
Apologies for raising a dup of question raised on Stackoverflow - http://stackoverflow.com/questions/42384352/springboot-1-5-1-upgrade-mvc-exception-handling-issues but given there was no helpful response on the question yet and given we are not able to fix it anyhow, we strongly belief it's a case of unexplained change in behaviour.

Kindly a take a look and advise.

In our application we use `@ControllerAdvice` and `@ExceptionHandler` to handle all kind of `Exceptions` and report those as `INTERNAL_SERVER_ERROR` as below:

    @ControllerAdvice
    public class ControllerExceptionHandler {
    
        @ResponseStatus(HttpStatus.INTERNAL_SERVER_ERROR)
        @ExceptionHandler(Exception.class)
        public String handleAllExceptions() {
            return "error";
        }
    }

Besides that, we also have a config class to define a handler that takes care of `NOT_FOUND` exceptions for the application using `ConfigurableEmbeddedServletContainer` bean to return a customised **pageNotFound** template as below:

    @Configuration
    public class WebConfig extends WebMvcConfigurerAdapter {
    
        @Override
        public void addViewControllers(ViewControllerRegistry registry) {
            registry.addViewController("/").setViewName("home");
            registry.addViewController("/404").setViewName("pageNotFound");
        }
    
        @Bean
        public EmbeddedServletContainerCustomizer containerCustomizer() {
            return (container) -> container.addErrorPages(new ErrorPage(HttpStatus.NOT_FOUND,
                                                                              "/404"));
        }

This has been working all fine till SBSP 1.4.X versions but since the date we migrated our app to SBSP 1.5.1, the `NOT_FOUND` handler is not working as expected and the app instead now returns the custom Error page template as defined in `handleAllExceptions` method as above even for `http-404`.

On further investigation, we removed the `@ControllerAdvice` class and the `error.html` template file in order to see if that lets the added NOT_FOUND handler come into play. But then the White-label Error Page for 404 is displayed instead of the customised pageNotFound template. 

So in summary it appears that, ConfigurableEmbeddedServletContainer bean configuration is failing to give the desired results (or at least the result it use to give prior to SBSP 1.5.X).

This may be a change in behaviour of the `ConfigurableEmbeddedServletContainer` configuration since 1.5.X but unexplained.

Any suggestions or guidance would be helpful to know the reason and a fix/workaround.

Thanks in adv!

**Comment by wilkinsona** [comment]
Let's keep things on Stack Overflow for now. We can re-open this if it turns out that a change is needed in Spring Boot. In the meantime, please make it easier for people to help you by updating your question on Atack Overflow with the recommended [minimal, complete, verifiable example](http://Stack Overflow.com/help/mcve) of the problem.

**Comment by ashwgupt** [comment]
@wilkinsona the question on stackoverflow is updated as per your suggestion.

**Comment by ashwgupt** [comment]
Here is the repo to recreate it now - https://github.com/ashwgupt/stackoverflow-42384352

**Comment by wilkinsona** [comment]
Thank you. That gave me what I needed to see what's going on. It's a bug with the same root cause as #8477 which I have just fixed.
